docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,315 @@
+
+    ```yaml
+    hibernation:
+      enabled: <bool>
+    ```
+
+  - Add `etcd.status.hibernatedAt`:  
+    This field conveys information about whether the cluster has been successfully hibernated after a reconciliation, and the time at which it entered hibernation. This field is cleared when the cluster is woken up from hibernation.
+
+    ```yaml
+    hibernatedAt: <hibernation-time>
+    ```
+
+  - Add `immutableSettings.retentionType` under `etcd.spec.backup.store`.
+- **ETCD Controller Logic:**
+  - When hibernation is requested, by changing `etcd.spec.hibernated.enabled` to `true`:
+    - The controller removes the ETCD client ports `2380` and `2379` from the `etcd-client` service, leaving only the etcd-backup-restore port `8080`. This stops ETCD client traffic.
+    - The controller creates a `EtcdOperatorTask` to trigger an on-demand full snapshot.
+      - On-demand full snapshot is successful: the controller does no additional handling.
+      - On-demand full snapshot fails: the controller triggers the creation of a `EtcdOperatorTask` for an on-demand compaction job that compacts the latest base full snapshot and the corresponding deltas.
+    - The controller scales in the ETCD cluster (i.e., sets `StatefulSet.spec.replicas` to zero).
+    - The controller creates the `EtcdSnapshotImmutabilityExtension` periodically if `etcd.spec.backup.store.immutableSettings.retentionType` is set to `"Bucket"`.
+
+- **`EtcdSnapshotImmutabilityExtension` specification:**
+  - Run `etcdbrctl extend-immutability --bucket-level-immutability` to extend the immutability of the latest snapshot.
+  - Run `etcdbrctl garbage-collect --garbage-collection-policy <garbage-collection-policy>` to garbage collect the snapshots that are created during the extension.
+
+- **EtcdOperatorTask Controller Logic:**
+  - The operator-tasks controller will react to the creation of the custom resource, and will deploy a job named `<etcd-name>-extend-immutability` which is the `EtcdSnapshotImmutabilityExtension` job.
+  - The controller also reports metrics regarding the `EtcdSnapshotImmutabilityExtension` job, which can be used to raise alerts for operators that immutability has not been extended.
+
+###### Advantages
+
+- **Minimal Operational Impact:** Does not alter the ETCD cluster's state during hibernation and respects the operator's intention to hibernate the cluster without unintended changes.
+- **Efficient Resource Utilization:** Only the latest snapshot is copied, limiting additional storage usage, and avoids the need to start an embedded ETCD instance.
+- **Automated Process:** The process of taking a full snapshot before hibernation and creating the `EtcdSnapshotImmutabilityExtension` is automated within the controller.
+
+###### Disadvantages
+
+- **Additional Complexity:** Requires updates to the etcd controller, introduction of the operator-tasks controller, and introduction of new etcdbrctl commands.
```

Comment:
Why do we see this as additional complexity, weren't we planning to anyways implement an operator-task controller?

Reply:
Additional complexity since it is a hard prerequisite; but you are right. This is the only way forward; will remove this.